File: data structures and algorithms/cpp/08_061.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    // Finds maximum product path in a grid
    int maxProductPath(vector<vector<int>>& grid) {
        // Check if grid is empty
        if (grid.empty() || grid[0].empty()) {
            return 0;
        }
        
        int rows = grid.size();
        int cols = grid[0].size();
        // DP arrays for max and min products
        vector<vector<long long>> maxDp(rows, 
                                vector<long long>(cols, 0));
        vector<vector<long long>> minDp(rows, 
                                vector<long long>(cols, 0));
        
        // Initialize the first cell
        maxDp[0][0] = minDp[0][0] = grid[0][0];
        
        // Initialize first row
        for (int j = 1; j < cols; j++) {
            maxDp[0][j] = maxDp[0][j-1] * grid[0][j];
            minDp[0][j] = minDp[0][j-1] * grid[0][j];
        }
        
        // Initialize first column
        for (int i = 1; i < rows; i++) {
            maxDp[i][0] = maxDp[i-1][0] * grid[i][0];
            minDp[i][0] = minDp[i-1][0] * grid[i][0];
        }
        
        // Fill the dp arrays
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                // For positive or zero values
                if (grid[i][j] >= 0) {
                    maxDp[i][j] = max(maxDp[i-1][j], 
                                    maxDp[i][j-1]) * grid[i][j];
                    minDp[i][j] = min(minDp[i-1][j], 
                                    minDp[i][j-1]) * grid[i][j];
                } 
                // For negative values
                else {
                    maxDp[i][j] = min(minDp[i-1][j], 
                                    minDp[i][j-1]) * grid[i][j];
                    minDp[i][j] = max(maxDp[i-1][j], 
                                    maxDp[i][j-1]) * grid[i][j];
                }
            }
        }
        
        // Return the maximum product at the bottom-right cell
        return maxDp[rows-1][cols-1];
    }
};

File: data structures and algorithms/python/08_061.py
def maxProductPath(grid):

    if not grid or not grid[0]:
        return 0
    
    rows, cols = len(grid), len(grid[0])
    max_dp = [[0 for _ in range(cols)] for _ in range(rows)]
    min_dp = [[0 for _ in range(cols)] for _ in range(rows)]
    
    # Initialize the first cell
    max_dp[0][0] = min_dp[0][0] = grid[0][0]
    
    # Initialize first row
    for j in range(1, cols):
        max_dp[0][j] = max_dp[0][j-1] * grid[0][j]
        min_dp[0][j] = min_dp[0][j-1] * grid[0][j]
    
    # Initialize first column
    for i in range(1, rows):
        max_dp[i][0] = max_dp[i-1][0] * grid[i][0]
        min_dp[i][0] = min_dp[i-1][0] * grid[i][0]
    
    # Fill the dp arrays
    for i in range(1, rows):
        for j in range(1, cols):
            if grid[i][j] >= 0:
                max_dp[i][j] = max(
                    max_dp[i-1][j], max_dp[i][j-1]
                ) * grid[i][j]
                min_dp[i][j] = min(
                    min_dp[i-1][j], min_dp[i][j-1]
                ) * grid[i][j]
            else:
                max_dp[i][j] = min(
                    min_dp[i-1][j], min_dp[i][j-1]
                ) * grid[i][j]
                min_dp[i][j] = max(
                    max_dp[i-1][j], max_dp[i][j-1]
                ) * grid[i][j]
    
    return max_dp[rows-1][cols-1]